### ISSUE TITLE
modelid GLEDOPTO unknown light model

### ISSUE THREAD

**Issue by b2un0** [ebaauw/homebridge-hue#244]
i use this new (?) led rgbw controller: https://www.amazon.de/gp/product/B077G23JBF/

and got this error after restart homebridge:

```
Dec 30 20:38:55 debian homebridge[6474]: [2017-12-30 20:38:55] [Hue] Philips hue: BSB002 bridge v1711151408, api v1.22.0
Dec 30 20:38:55 debian homebridge[6474]: [2017-12-30 20:38:55] [Hue] Philips hue: /lights/8: warning: using default colour temperature range for unknown light model {"state":{"on":true,"bri":253,"hue":7424,"sat":20,"effect":"none","xy":[0.5213,0.4635],"ct":470,"alert":"select","colormode":"xy","mode":"homeautomation","reachable":true},"swupdate":{"state":"notupdatable","lastinstall":null},"type":"Extended color light","name":"Bett","modelid":"GLEDOPTO","manufacturername":"GLEDOPTO","capabilities":{"streaming":{"renderer":false,"proxy":false}},"uniqueid":"00:12:4b:00:13:0a:5c:90-0b","swversion":"1.0.2"}
Dec 30 20:38:55 debian homebridge[6474]: [2017-12-30 20:38:55] [Hue] Philips hue: /lights/8: warning: using default colour gamut for unknown light model {"state":{"on":true,"bri":253,"hue":7424,"sat":20,"effect":"none","xy":[0.5213,0.4635],"ct":470,"alert":"select","colormode":"xy","mode":"homeautomation","reachable":true},"swupdate":{"state":"notupdatable","lastinstall":null},"type":"Extended color light","name":"Bett","modelid":"GLEDOPTO","manufacturername":"GLEDOPTO","capabilities":{"streaming":{"renderer":false,"proxy":false}},"uniqueid":"00:12:4b:00:13:0a:5c:90-0b","swversion":"1.0.2"}
Dec 30 20:38:55 debian homebridge[6474]: [2017-12-30 20:38:55] [Hue] Philips hue: 3 accessories
Dec 30 20:38:55 debian homebridge[6474]: [2017-12-30 20:38:55] [Hue] Initializing platform accessory 'Philips hue'...
Dec 30 20:38:55 debian homebridge[6474]: [2017-12-30 20:38:55] [Hue] Philips hue: 3 services
Dec 30 20:38:55 debian homebridge[6474]: [2017-12-30 20:38:55] [Hue] Initializing platform accessory 'Flur'...
Dec 30 20:38:55 debian homebridge[6474]: [2017-12-30 20:38:55] [Hue] Initializing platform accessory 'Bett'...
Dec 30 20:38:55 debian homebridge[6474]: [2017-12-30 20:38:55] [Hue] Error: Cannot add a Service with the same UUID '00000043-0000-1000-8000-0026BB765291' as another Service in this Accessory without also defining a unique 'subtype' property
```
i use a standard SMD 5050 RGBW LED Stripe with this Controller, but the Controller register 2 devices into hue bridge, a color and white device.

any suggestions how to fix this via blacklist?

**Comment by ebaauw** [comment]
The controller probably has two ZigBee endpoints, one for the RGB channels and one for the W channel.  The Hue bridge (and deCONZ gateway) exposes this is two resources, in your case `/lights/8` as an _Extended colour light_, and another one, probably as a _Dimmable light_.  If you run `homebridge -D`, homebridge-hue will log debug messages about your bridge configuration, otherwise you need to check this using the Hue API.  You'll recognise the resource by its `uniqueid`, since it will start with the same MAC address, but with a different endpoint `"00:12:4b:00:13:0a:5c:90-xx"`.  You would need to blacklist this resource.

More structurally, I'll add the controller to homebridge-hue's list of known lights, indicating that it has multiple endpoints.  That will get rid of the _same UUID_ error.  To get rid of the warnings: could you find out what colour range and colour temperature range the controller supports?  I still need to build this discovery into homebridge-hue.  For now, I've got a bash command, `ph_light_values`, doing that included in [ph.sh](https://github.com/ebaauw/ph.sh).

**Comment by b2un0** [comment]
thanks for your reply, i used your tool and got this:

```
root@debian:/opt/ph.sh# ph_light_values 10
192.168.177.40: /lights/10: GLEDOPTO GLEDOPTO Extended color light "Extended color light 1"
192.168.177.40: /lights/10: ct: cool ... 153
192.168.177.40: /lights/10: ct: warm ... 500
192.168.177.40: /lights/10: xy: red ............................ [0.3125,0.3333]
192.168.177.40: /lights/10: xy: green ..... [0.3125,0.3333]
192.168.177.40: /lights/10: xy: blue ............................................................... [0.3125,0.3333]
192.168.177.40: /lights/10: done
{
  "manufacturer": "GLEDOPTO",
  "modelid": "GLEDOPTO",
  "type": "Extended color light",
  "bri": true,
  "ct": {
    "min": 153,
    "max": 500
  },
  "xy": {
    "r": [
      0.3125,
      0.3333
    ],
    "g": [
      0.3125,
      0.3333
    ],
    "b": [
      0.3125,
      0.3333
    ]
  }
}
```

2nd device
```
root@debian:/opt/ph.sh# ph_light_values 11
192.168.177.40: /lights/11: GLEDOPTO GLEDOPTO Dimmable light "Dimmable light 1"
192.168.177.40: /lights/11: done
{
  "manufacturer": "GLEDOPTO",
  "modelid": "GLEDOPTO",
  "type": "Dimmable light",
  "bri": true
}
```

**Comment by ebaauw** [comment]
That doesn’t look good.  The controller reports the same `xy` values for red, green, and blue.  Did the strip change colour while running `ph_light_values`?  Can you change its colour through the Hue API using `xy`?  Using `hue`/`sat`?  Does the Hue app reflect the correct colour?

**Comment by ebaauw** [comment]
GLEDOPTO should be supported by v0.5.55, although still with default gamut (but no more warning).

**Comment by b2un0** [comment]
> Did the strip change colour while running ph_light_values?
not all time, if i use it with Philips LEDs everything changes correctly 
the GLEDOPTO lags every time

> Can you change its colour through the Hue API using xy? Using hue/sat?
Yes, but with a small delay

> Does the Hue app reflect the correct colour?
Yes

> GLEDOPTO should be supported by v0.5.55
thank you, i will test this

**Comment by ebaauw** [comment]
> not all time, if i use it with Philips LEDs everything changes correctly the GLEDOPTO lags every time

I suspect the GLEDOPTO might not support setting colour though `xy`.

Could you double-check:
```
ph_put /lights/10/state '{"on": true, "ct": 153}'
sleep 1
ph_get /lights/10/state
ph_put /lights/10/state '{"xy": [1, 0]}'
ph_get /lights/10/state
sleep 300
ph_get /lights/10/state
```
The `ct` is to set the light to white; the `xy` should turn it red(-ish) immediately (but I suspect it doesn't).  The bridge should report `xy` as `[1,0000, 0.0000]` after that.  After a while (once the Hue bridge has polled the light) it should have changed `xy` to reflect the actual colour.  `[0.3125,0.3333]` corresponds to white(-ish), corresponding to `ct` 153.

What about:
```
ph_put /lights/10/state '{"on": true, "ct": 153}'
sleep 1
ph_put /lights/10/state '{"hue": 0, "sat": 100}'
```
Now, the light should turn red(-ish) immediately.

And what about:
```
ph_put /lights/10/state '{"on": true, "ct": 153}'
sleep 1
ph_put /lights/10/state '{"xy": [1, 0], "transitiontime": 0}'
```

**Comment by b2un0** [comment]
```
root@debian:/opt/ph.sh# ph_put /lights/10/state '{"on": true, "ct": 153}'
root@debian:/opt/ph.sh# sleep 1
root@debian:/opt/ph.sh# ph_get /lights/10/state
{
  "on": true,
  "bri": 244,
  "hue": 5632,
  "sat": 15,
  "effect": "none",
  "xy": [
    0.5777,
    0.4119
  ],
  "ct": 153,
  "alert": "none",
  "colormode": "ct",
  "mode": "homeautomation",
  "reachable": true
}
root@debian:/opt/ph.sh# ph_put /lights/10/state '{"xy": [1, 0]}'
root@debian:/opt/ph.sh# ph_get /lights/10/state
{
  "on": true,
  "bri": 244,
  "hue": 5632,
  "sat": 15,
  "effect": "none",
  "xy": [
    1.0000,
    0.0000
  ],
  "ct": 153,
  "alert": "none",
  "colormode": "xy",
  "mode": "homeautomation",
  "reachable": true
}
root@debian:/opt/ph.sh# sleep 300
root@debian:/opt/ph.sh# ph_get /lights/10/state
{
  "on": true,
  "bri": 244,
  "hue": 5632,
  "sat": 15,
  "effect": "none",
  "xy": [
    0.5777,
    0.4119
  ],
  "ct": 153,
  "alert": "none",
  "colormode": "ct",
  "mode": "homeautomation",
  "reachable": true
}
```

im currently not at home and fire the commands over ssh on my NAS so i cant see whats going on, but the Hue App shows an "cyan blue" state.

**Comment by ebaauw** [comment]
OK, you see that `colormode` is reset to `ct` in the first run, after 5 minutes.  This confirms my suspicion that the light doesn't support setting `xy`.  You're not running any automation that would change the light, are you?

> but the Hue App shows an "cyan blue" state

Could be the blue-ish white colour temperature of 153 (~6500K), or default `xy` values of `[0, 0]`.

Could you get the state and/or check the Hue app right after and ~5 mins after:
```
ph_put /lights/10/state '{"hue": 0, "sat": 100}'
ph_get /lights/10/state
sleep 300
ph_get /lights/10/state
```
I'm hoping the Hue app reports red(-ish), at least after ~5 mins.

**Comment by b2un0** [comment]
i wil test this tomorrow.

**Comment by Oggymator** [comment]
Hi,
I'm one of the employee of the company that sell and produce the products so maybe I can try to help a bit.
Is it more easy to send Erik some samples for test?
I saw you came from Netherland and we from Germany, so it is a short way for shipping.
Thanks b2un0 for feedback on amazon, so I can try to support.
Erik, you can contact me direct if you like tobias.ebert@led-trading.de

**Comment by ebaauw** [comment]
Thanks, Thobias, that would be helpful.  I’ll be more than happy to test drive the controller and will contact you by email.

**Comment by ebaauw** [comment]
Today, I received the samples, thanks @Oggymator!

I first paired them to deCONZ and see what they look like.  The RGBW, RGB+CCT, and Flood light all report the same Manufacturer and _Model ID_, `GLEDOPTO`, _Date Code_ ,`201312061`, and _SW Build ID_, `1.0.2`.  The WW/CW reports a different _Date Code_, `20170724`, and _SW Build ID_, `1.0.3`.

As expected, the RGBW and the Flood light seem to have the same driver module.  It provides three endpoints, 0x0A, a _ZLL Dimmable Light_, 0x0B, a ZLL _Extended Color Light_, and 0x0D, with device ID 0xE15E for the (client and server) _ZLL Commissioning_ clusters.  No OTAU, no attribute reporting, but otherwise everything you'd expect.  The _Dimmable Light_ drives the white channel, the _Extended Color Light_ drives the RGB channels.  Both endpoints support 16 groups and 16 scenes.  The lights turn on at the previous level.  When powered on after a brief power interruption, they seem to turn on at the previous level as well.  Need to check what they do overnight.

The _Extended Color Light_ endpoint supports _Hue/Saturation_, _Enhanced Hue/Saturation_, _X/Y_, and _Color Temperature_.  Both _Color Mode_ and _Enhanced Color Mode_ reflect the correct value.  When setting _(Enhanced) Hue/Saturation_, the light updates _Current X_ and _Current Y_; when setting _X/Y_, it updates _(Enhanced) Hue_ and _Saturation_.  The reflect the colours I set from HomeKit fairly accurately.  The flood light has really brilliant colours, comparable to the Hue Bloom.

 _Color Temperature_ seems a world in itself.  The lights don't report ctmin and ctmax.  The RGBW reflects the colour temperature from 153 to 500 fairly accurate; the Flood light turns red at the higher mired values.  When setting _Colour Temperature_ to 1 or 1000, the lights don't report the actual value.  I think it's best to ignore the colour temperature when exposing these to HomeKit, and just treat them as a _Color Light_.  There's no way to distinguish the RGBW and the Flood light at the ZigBee level.

The RGB+CCT works differently from what I've seen before (I don't have any Hue LightStrip+).  It exposes only two endpoints, the 0x0B ZLL _Extended Color Light_ and the 0x0D for ZLL Commissioning, so it's a single light resource in the REST API.  The 0x0B endpoint looks the same as for the RGBW, but behaves differently.  Setting _Color Temperature_ turns off the RGB channels, setting _(Enhanced) Hue/Saturation_ or _X/Y_ turns off the CW and WW channels.  I've not been able to power all five channels at the same time.  Both colour and colour temperature reflect the setting from HomeKit.  When setting `ct` to out of range values, the light reports the actual value.  It supports the full 153-500 range (the warm LED turns off below 160 and the cool LED turns off at 500).

The WW/CW looks very much like the RGB+CCT, except, of course, that it only supports _Color Temperature_.

In summary: no issues supporting these in homebridge-hue.

All lights seem to dislike transitiontime, this could be what's causing the funny behaviour with the Hue bridge (which adds a default transitiontime of 4to ZigBee commands).  I can force homebridge-hue to send transitiontime of 0 when connected to a Hue bridge if this is really an issue.  Need to do some testing and maybe some sniffing to find out.  More to follow...

**Comment by Oggymator** [comment]
Hi, thanks for the very long first feedback.

For RGBW you can ignore color temperature.

RGB+CCT should be like the Philips Hue Lightstripe+ this use also RGB and color temperature.

OTAU is at the moment not supported, that is correct. We will update the hardware in this way for a later version.

For what is the transitiontime?
Do you if it is an important thing and we have to improve the software in this part?

What do you think about the 5in1 stripe?
I like it, because every LED can RGB+Warmwhite+Coldwhite :)

**Comment by ebaauw** [comment]
> For RGBW you can ignore color temperature.

It's going to be a (small) challenge recognising the RGBW.  The resource on the Hue bridge (or on deCONZ) for the _Enhanced Color Light_ endpoint looks exactly the same as the RGB+CCT resource.  I have to look for the matching other resource (for the _Dimmable Light_) to recognise the difference.

> RGB+CCT should be like the Philips Hue Lightstripe+ this use also RGB and color temperature.

I don't have any gen3 Hue Extended Color Lights, like the Hue LightStrips Plus, so I cannot check.  It would seem the `xy` and `hue`/`sat` aren't updated when setting `ct` and v.v.; they are on the gen2 Hue Extended Color Lights.  I need to do some more testing.

> For what is the transitiontime?

To change colour (temperature) or brightness slowly.  E.g. for my wakeup alarm, I turn my bedroom lights from 0 to full brightness in 10 minutes using a transitiontime of 6000 (deciseconds).  I need to do some more testing here as well.

> What do you think about the 5in1 stripe?

Very cool!  Makes me almost want to break open my (gen2) Hue Extended Color Lights to see what LEDs are in there.  I think they use only RGB, but with different colours than usual, so they can better emulate the whites, at the expense of the greens and blues.  
Is does make the RGBW controller seem old-fashioned, though.

I did find the issue with `ph_light_values` and the colours sometimes not matching.  It's a brilliant case of different manufacturers taking different interpretations of the ZigBee standards.  I need to do some sniffer to confirm the cause, but I know how to prevent the issues from happening.  I'll do a write-up tomorrow.

**Comment by ebaauw** [comment]
Did some extensive testing.  There's some subtle differences in how various manufacturers deal with the ZigBee standards, causing the issues with `ph_light_values` and with the unexpected colours mentioned above.  I've found issues in the GLEDOPTO firmware and in the Hue bridge firmware, but nothing that cannot be worked around.

In ZigBee, the values for `xy` are uint16.  The ZCL spec states that the maximum value for these is 65279 (0xfeff).  The minimum value is 0, but for `y`, this might cause a division by zero error when converting the colour to `hue`/`sat`.  I ran into this issue in homebridge-hue.  The firmware in the GLEDOPTO as well as in the IKEA Trådfri CWS seems to run into the same issue: the light goes into some sort of error (the GLEDOPTO turn green-ish; the Trådfri turns blue-ish and flickers when setting brightness afterwards). 

From the ZigBee Cluster Library Specification:
> _CurrentX_ Attribute
The _CurrentX_ attribute contains the current value of the normalized chromaticity value x, as defined in the CIE xyY Color Space. It is updated as fast as practical during commands that change the color.
The value of x SHALL be related to the _CurrentX_ attribute by the relationship
x = _CurrentX_ / 65536 (_CurrentX_ in the range 0 to 65279 inclusive)
_CurrentY_ Attribute
The _CurrentY_ attribute contains the current value of the normalized chromaticity value y, as defined in the CIE xyY Color Space. It is updated as fast as practical during commands that change the color.
The value of y SHALL be related to the _CurrentY_ attribute by the relationship
y = _CurrentY_ / 65536 (_CurrentY_ in the range 0 to 65279 inclusive)

The Hue bridge divides the ZigBee values by 65536, to get the REST API `xy` values from 0 to 0.9999.  When setting `xy` to `[0, 1]` through the Hue API, it multiplies the values by 65536, e.g. the bridge issues a _Move to colour_ command with parameters _X_: 0x0000, _Y_: 0xFFFF, and _Transition time_: 4.  I think this is a bug: the bridge should send _Y_: 0xFEFF (the maximum allowed value).  deCONZ actually maximises the values to 65279, but has another bug: it divides the ZigBee values by 65279, leading to slightly different REST API values, from 0 to 0.9961 (see https://github.com/dresden-elektronik/deconz-rest-plugin/issues/30 and https://github.com/dresden-elektronik/deconz-rest-plugin/issues/195#issuecomment-334960340).

Different lights support different colours.  Philips actually uses three different colour gamuts: different Hue lights support different colours.  The Hue bridge knows which Hue light supports which colour gamut.  When setting `xy` of a Hue light, the Hue bridge minimises/maximises the values, to fall within the gamut the light supports, so Hue bridge always reports the actual light colour.  For non-Hue lights, and for groups, the Hue bridge just reports the given values, until the light is next polled.

The Hue lights are happy receiving values they don't support, even values above 65279.  They turn to the closest colour they support anyway and report back their actual value.  The innr works just like that (see #152).  The Trådfri colour light accepts any value (except an _Y_ value of 0), but always echoes back the given value (see https://github.com/dresden-elektronik/deconz-rest-plugin/issues/218).

The GLEDOPTO rightly returns an error when receiving values above 65279 and stays on the current colour.  It correctly reports the _previous_ values of _X_ and _Y_ (which do _not_ reflect the actual colour).  I think the JAIWEN probably does the same (see #171).  The Hue bridge gracefully ignores the error returned by the GLEDOPTO and the REST API client is left in the dark (or rather the previous colour) about what happened.  As the REST API probably returns when the ZigBee command is queued, there's no way for the bridge to return the error.  Luckily the GLEDOPTO reports the unchanged values the next time the bridge polls the light.

I don't think this is a common issue in real-life applications, because they compute `xy` from RGB or HSV values, so the result (usually?) falls within the valid range.  However, when testing what colours a light supports, `ph_light_values` sends `xy` values of `[1, 0.0001]`, `[0, 1]` and `[0, 0.0001]`, for red, green and blue, which on the Hue bridge translates to 65535.

Concluding (read: notes to self):
- homebridge-hue should use a default colour gamut of:
   ```json
   {
     "r": [0.9961, 0.0001],
     "g": [0, 0.9961],
     "b": [0, 0.0001]
   }
   ```
- `ph_light_values` should use the same for testing a light;
- deCONZ should divide/multiply by 65535, but still maximise the values to 65279.

**Comment by ebaauw** [comment]
Some more fun facts:

1. The Flood doesn't turn on when receiving a _Move to level (with On/Off)_ command with _Level_ set to the value the light had when turned off.  In REST API speak: setting e.g. `{"bri": 123}`, then `{"on":false}`, then `{"on":true, "bri": 123}` doesn't turn on the Flood back on.  Setting `{"on":true, "bri":124}` (or any value for `bri` except 0 and 123) does turn on the Flood.  The RGBW doesn't have this issue, so there's definitely some differences in the ZigBee hardware and/or firmware.  
This wouldn't be an issue when using HomeKit interactively: the HomeKit app would still be on the previous Brightness setting, and the only way to generate a combined `on` and `bri` request is by moving the slider.  However, this will be an issue when recalling HomeKit scenes that include Brightness.  homebridge-hue can probably add 1 to the `bri` value when the request includes `bri` and `"on": true` and the light is currently off and at the requested `bri` level;
2. The Flood doesn't transition smoothly.  With the Hue bridge default transition time of 4 (0.4 seconds), there's no (noticeable) transition.  With longer transition times, e.g. 40 (4 seconds), the Flood transitions using noticeable discrete steps;
3. All GLEDOPTO lights seem to store the current state in non-volatile memory.  On power up, they turn on, but with the previous brightness and colour.  For the RGBW and Flood this happens on both endpoints.
4. There's no way for homebridge-hue to distinguish between the RGBW and the Flood, other than probing the light for issue 1.  The RGB+CCT can be recognised because it doesn't have an 0x0a _Dimmable Light_ endpoint, but that's kinda hard to do for homebridge-hue, as I currently go linearly through the light resources, and there's no guarantee which endpoint appears first.  The WW/CW can be recognised easily because it's a _Color temperature light_ instead of an _Extended colour light_.  
This makes it hard to implement model-specific workarounds, like for 1) and to suppress colour temperature for the RGBW and Flood.  In future firmware, please use different values for _Model identifier_.

**Comment by Oggymator** [comment]
RGB+CCT and RGBW/Floodlight has the same Model indentifier?

RGBW controller and Floodlight use the same software and ZLL module at the moment, so it is ok if you cannot distinguish.

**Comment by ebaauw** [comment]
v0.5.66 supports CC/WC, RGBW, and RGB+CCT ballasts and Flood and Garden lights.
